### PR TITLE
FLINK-23454 The upstream sends the buffer of the right size for unicast case

### PR DESCRIPTION
**Pull request by akalash**
## What is the purpose of the change

*This PR implements the reaction on NewBufferSize. It changes the size of output buffers corresponded to the received information.*

## Brief change log

  - *Added the support of changing the max capacity to BufferBuilder*
  - *Subpartition is able to notify the desirable buffer size for input buffer consumer.*
  - *Notifying the subpartitions about the new received buffer size.*
  - *LocalInputChannel notifying the subpartition about the new buffer size directly. *
  - *ResultPartition changes the BufferBuilder size according to the desirable buffer size of each subpartition. *

## Verifying this change

  - *Added unit tests for each classes which behaviour was changed*
## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java [hunk 1]
```diff
@@ -450,6 +454,16 @@ public int getNumberOfQueuedBuffers() {
         }
     }
 
+    @Override
+    public void bufferSize(int desirableNewBufferSize) {
+        if (desirableNewBufferSize < 0) {
+            return;
+        }
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:461]: Why ignore negative numbers? Should we throw `IllegalArgumentException`?

**Review by pnowojski — COMMENTED** [review]
I think change looks mostly good to me. 

Have you checked benchmark-request for this PR?

**Comment by akalash** [comment]
@flinkbot run azure

**Comment by akalash** [comment]
@pnowojski , tests finally are green. So please pay attention to the first(https://github.com/apache/flink/pull/16726/commits/62e69286ad4a4a2367bac131fed2cbe0b71337f5) and the last(https://github.com/apache/flink/pull/16726/commits/8b6e33923bfbb843e425dc87d65520e595c5fea8) commits which were changes recently. Also, benchmarks look good:

- http://codespeed.dak8s.net:8080/job/flink-benchmark-request/410/
- http://codespeed.dak8s.net:8080/job/flink-benchmark-request/429/

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java [hunk 2]
```diff
@@ -396,12 +397,13 @@ protected StreamTask(
 
         environment.getMetricGroup().getIOMetricGroup().setEnableBusyTime(true);
         this.throughputCalculator = environment.getThroughputCalculator();
-        this.bufferDebloatPeriod = getTaskConfiguration().get(BUFFER_DEBLOAT_PERIOD).toMillis();
+        Configuration taskManagerConf = environment.getTaskManagerInfo().getConfiguration();
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:400]: Can you create a separate jira ticket for this bug fix (as this bug is already on the master)? This fix can stay in this PR, but can you adjust it's commit message to match the crated ticket?

**Review by pnowojski — APPROVED** [review]
LGTM % one comment

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java (hunk 2 above)
Comment (akalash) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:400]: done. https://issues.apache.org/jira/browse/FLINK-23726

**Pull request merged**